api/lib/infrastructure/utils/domain-builder.js

```diff
@@ -0,0 +1,51 @@
+const _ = require('lodash');
+const Bookshelf = require('../bookshelf');
+const Models = require('../../domain/models');
+
+const belongsToKeys = _.keys(Models).map((key) => _.lowerFirst(key));
+const hasManyKeys = _.keys(Models).map((key) => _.lowerFirst(key) + 's');
+
+module.exports = {
+  buildDomainObjects,
+  buildDomainObject,
+};
+
+function buildDomainObjects(bookshelfObjects) {
+  if (bookshelfObjects.length === 0) {
+    return [];
+  }
+
+  const modelName = _getModelName(bookshelfObjects[0]);
+
+  return bookshelfObjects.map(
+    (bookshelfObject) => _buildDomainObject(new Models[modelName], bookshelfObject.toJSON())
+  );
+}
+
+function buildDomainObject(bookshelfObject) {
+  const modelName = _getModelName(bookshelfObject);
+
+  return _buildDomainObject(new Models[modelName], bookshelfObject.toJSON());
+}
+
+function _getModelName(bookshelfObject) {
+  return _.findKey(Bookshelf._models, (BookshelfModel) => {
+    return bookshelfObject instanceof BookshelfModel;
+  });
+}
+
+function _buildDomainObject(domainObject, bookshelfObject) {
+  return _.mapValues(domainObject, (value, key) => {
+    if (belongsToKeys.includes(key) && bookshelfObject[key] instanceof Object) {
+      return _buildDomainObject(new Models[_.upperFirst(key)], bookshelfObject[key]);
+    }
+
+    if (hasManyKeys.includes(key) && Array.isArray(bookshelfObject[key])) {
+      return bookshelfObject[key].map(
+        (bookshelfObject) => _buildDomainObject(new Models[_.upperFirst(key).slice(0, -1)], bookshelfObject)
+      );
+    }
+
+    return bookshelfObject[key];
+  });
```

Comment:
Je pense qu’ici on a un problème de branchement conditionnelles : impossible de lire à première vue quels sont tous les cas.
Symboliquement, on a les conditions suivantes:  `A AND B`, `C AND D`, ainsi que `NON (A AND B) AND (C AND D)`. Que ce passe-t-il lorsqu’on a un par exemple : `NON A AND B` ? C’est à dire un objet qui n’appartient pas à la liste des propriétés possibles ? 
Soit ça n’arrivera jamais auquel cas y’a un problème de clarté je pense qu’il faut refactored, soit ça peut arriver et causer des pb, auquel cas 
Faut le documenter dans les tests. Je pense que si on veut faire du code intelligent et aussi générique que tout ce qui est entrepris dans cette PR (ce que je trouve vrm cool :)), faut être bullet-proof sur les branchements possibles.

Reply:
En effet il manque un test, je le rajoute.